File: leetcode/383.ransom-note/ts/canconstruct.ts
// M + N, 1
const canConstruct = (ransomNote: string, magazine: string): boolean => {
    const requiredCharacterCounts = Array.from({ length: 26 }, () => 0)

    for (const ch of ransomNote) requiredCharacterCounts[dist(ch)]++

    for (const ch of magazine) requiredCharacterCounts[dist(ch)]--

    return !requiredCharacterCounts.some(v => v > 0)
};

const dist = (ch: string): number => {
    return ch.charCodeAt(0) - "a".charCodeAt(0)
}

export default canConstruct

File: leetcode/383.ransom-note/cpp/canconstruct.cpp
#include <string>
#include <array>

class Solution
{
public:
    bool canConstruct(std::string ransomNote, std::string magazine)
    {
        std::array<int, 26> arr{};

        for (auto it = ransomNote.begin(); it != ransomNote.end(); ++it)
            arr[dist(*it)]++;

        for (auto it = magazine.begin(); it != magazine.end(); ++it)
            arr[dist(*it)]--;

        return std::find_if(arr.begin(), arr.end(), [](int v)
                            { return v > 0; }) == arr.end();
    }

private:
    int dist(char ch)
    {
        return (int)(ch - 'a');
    }
};
